hw5/csvt32745/matrix.hpp

```diff
@@ -0,0 +1,357 @@
+                vec.push_back(Matrix(*this, r, c, size_r, size_block));
+            }
+            // remaining col
+            if(m_ncol - c > 0){
+                vec.push_back(Matrix(*this, r, c, size_r, m_ncol - c));
+            }
+            ans.push_back(vec);
+        }
+
+        return ans;
+    }
+
+    std::string reprString() const{
+        std::stringstream stm;
+        size_t idx = 0;
+        for(size_t i = 0; i < m_nrow; ++i){
+            for(size_t j = 0; j < m_ncol; ++j){
+                stm << m_data[idx] << " ";
+                idx++;
+            }
+            stm << std::endl;
+        }
+        return stm.str();
+    }
+
+    size_t index(size_t row, size_t col) const{
+        return row*m_ncol + col;
+    }
+
+    void reset_data(size_t nrow, size_t ncol) {
+        if(m_data) delete[] m_data;
+
+        const size_t size = nrow * ncol;
+        if(size > 0){
+            m_data = new double[size]{0};
+        } 
+        else m_data = nullptr;
+
+        m_nrow = nrow;
+        m_ncol = ncol;
+    }
+
+    private:
+    size_t m_nrow = 0;
+    size_t m_ncol = 0;
+    double* m_data = nullptr;
+};
+
+
+/* ========================
+       Multiplication
+   ========================*/
+
+Matrix multiply_naive(const Matrix& a, const Matrix& b)
+{
+    auto ans = Matrix(a.nrow(), b.ncol());
+    for(size_t r = 0; r < ans.nrow(); r++){
+        for(size_t c = 0; c < ans.ncol(); c++){
+            double s = 0;
+            for(size_t i = 0; i < a.ncol(); i++){
+                s += a(r, i) * b(i, c);
+            }
+            ans(r, c) = s;
+        }
+    }
+    return ans;
+}
+
+Matrix multiply_mkl(const Matrix& a, const Matrix& b)
+{
+
+    size_t m = a.nrow();
+    size_t k = a.ncol();
+    size_t n = b.ncol();
+    Matrix c = Matrix(m, n);
+    
+    cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans,
+    m, n, k, 1,
+    a.data(), k,
+    b.data(), n,
+    0, c.data(), n);
+
+    return c;
+}
+
+
+
+Matrix multiply_tile(const Matrix& a, const Matrix& b, size_t tile_size)
+{
+    // 2D Vec
+    auto A = a.split_block(tile_size);
+    auto B = b.split_block(tile_size);
+    // std::cout << A.size() << ", " << A[0].size() << "\n";
+    // std::cout << B.size() << ", " << B[0].size() << "\n";
+    auto ans = Matrix(a.nrow(), b.ncol());
+
+    size_t m = A.size();
+    size_t k = B.size();
+    size_t n = B[0].size();
+
+    // Block Multiplication
+    for(size_t r = 0, row_id = 0; r < m; r++){
+        size_t mm = A[r][0].nrow();
+        
+        for(size_t c = 0, col_id = 0; c < n; c++){
+            size_t nn = B[0][c].ncol();
+            Matrix s(mm, nn);
+            
+            for(size_t i = 0; i < k; i++){
+                s += multiply_naive(A[r][i], B[i][c]);
+            }
+            ans.set_block(s, row_id, col_id, mm, nn);
+            col_id += nn;
+        }
+        row_id += mm;
+    }
+    return ans;
+}
+
+// PYBIND11_MODULE(_matrix, m){

```

Comment:
why do you comment on all of this part?

Reply:
I thought `../mod.cpp` would do the pybind work for us

Reply:
I see. I think it's a good habit to clean useless code rather than commenting on them. It will make the reviewer confused.

Reply:
Got it, thank you so much.
It's due to my laziness cause I've already known that I might use the commented pybind part in hw6.